drivers/char/rust_example.rs

```diff
@@ -51,23 +53,120 @@ module! {
     },
 }
 
-struct RustFile;
+const MAX_TOKENS: usize = 3;
 
-impl FileOperations for RustFile {
+struct SharedStateInner {
+    token_count: usize,
+}
+
+struct SharedState {
+    state_changed: CondVar,
+    inner: Mutex<SharedStateInner>,
+}
+
+impl SharedState {
+    fn try_new() -> KernelResult<Arc<Self>> {
+        let state = Arc::try_new(Self {
+            // SAFETY: `condvar_init!` is called below.
+            state_changed: unsafe { CondVar::new() },
+            // SAFETY: `mutex_init!` is called below.
+            inner: unsafe { Mutex::new(SharedStateInner { token_count: 0 }) },
+        })?;
+        // SAFETY: `state_changed` is pinned behind `Arc`.
+        let state_changed = unsafe { Pin::new_unchecked(&state.state_changed) };
+        kernel::condvar_init!(state_changed, "SharedState::state_changed");
+        // SAFETY: `inner` is pinned behind `Arc`.
+        let inner = unsafe { Pin::new_unchecked(&state.inner) };
+        kernel::mutex_init!(inner, "SharedState::inner");
+        Ok(state)
+    }
+}
+
+struct Token {
+    shared: Arc<SharedState>,
+}
+
+impl FileOpener<Arc<SharedState>> for Token {
+    fn open(shared: &Arc<SharedState>) -> KernelResult<Self::Wrapper> {
+        Ok(Box::try_new(Self {
+            shared: shared.clone(),
+        })?)
+    }
+}
+
+impl FileOperations for Token {
     type Wrapper = Box<Self>;
 
-    kernel::declare_file_operations!();
+    kernel::declare_file_operations!(read, write);
+
+    fn read(&self, _: &File, data: &mut UserSlicePtrWriter, offset: u64) -> KernelResult<usize> {
+        // Succeed if the caller doesn't provide a buffer or if not at the start.
+        if data.is_empty() || offset != 0 {
+            return Ok(0);
+        }
+
+        {
+            let mut inner = self.shared.inner.lock();
+
+            // Wait until we are allowed to decrement the token count or a signal arrives.
+            while inner.token_count == 0 {
+                if self.shared.state_changed.wait(&mut inner) {
+                    return Err(Error::EINTR);
+                }
+            }
+
+            // Consume a token.
+            inner.token_count -= 1;
+        }
+
+        // Notify a possible writer waiting.
+        self.shared.state_changed.notify_all();
+
+        // Write a one-byte 1 to the reader.
+        data.write_slice(&[1u8; 1])?;
+        Ok(1)
```

Comment:
Side-note (no need to change anything for this): if `write_slice` returned the written size, then one could forward it in cases like this:

```suggestion
        data.write_slice(&[1u8; 1])
```

Reply:
I think returning the number of bytes written makes sense if it can ever be different from the expected value on success. For example, with stream sockets, a `write` of 100 bytes may return 50 if the send buffer fills up; this is still considered success and I think it makes sense.

In the case at hand, I think it's more of an all-or-nothing situation: we either write everything successfully or fail with `EFAULT`. So I think it's preferable to not return anything, otherwise callers may think that it could return less, which may lead them to write extra code unnecessarily (to handle such cases).

Reply:
Yeah, I am not a fan of returning things that are already known either.

Some people like to chain things though, specially if they come from C (e.g. `memcpy`, `memset`, `strcpy`...). But given the equivalents in Rust (`std::ptr::copy*`, `std::ptr::write_bytes`...) do not return anything, I guess we can say it is not idiomatic.

I will write it in the "coding guidelines" since I assume it will come up from time to time.